### ISSUE TITLE
Saving test parameters as a default parameter

### ISSUE THREAD

**Issue by YashUoM** [TestCentric/testcentric-gui#703]
Is there a way we can save the test parameter to have them even after closing and re-opening the exe, so that we do not need to enter test parameters every time we open the exe?

**Comment by CharliePoole** [comment]
Do you mean the parameters, which are available to your tests through TestContext?

**Comment by CharliePoole** [comment]
@YashUoM Waiting for you to clarify. :smiley:

**Comment by YashUoM** [comment]
@CharliePoole Yes TestContext parameters. Sorry about the late clarification

**Comment by CharliePoole** [comment]
Currently, there is no way to do this.

Since those parameters are stored in the TestPackage, I see this as a part of issue #3. We'll keep this alive to remind us that saving TestPackages has to include any parameters provided. At the moment, I'm afraid I don't have a schedule for finishing that issue but it will be incorporated in the 2.0 release.

**Comment by YashUoM** [comment]
Is this possible to do with the current code base? I'm trying to implement this and have around 60% success by keeping the test parameters in the user settings. However currently having issues with reloading tests with test parameters. A tip would be really nice. :-)

**Comment by CharliePoole** [comment]
I don't see any way to do it short of modifying TestCentric to save and restore these for each package. That's a fairly large change and I'm afraid my time is currently devoted to other issues related to the 2.0 release. If you are interested in working on the modification, let me know and I'll help you get set up to do that.

**Comment by YashUoM** [comment]
Well I'm currently working on this and currently stuck at how I can reload tests with test parameters. If you can help me set up this, it would be great

**Comment by CharliePoole** [comment]
What have you done so far? Are you building TC from source?

**Comment by YashUoM** [comment]
What I did so far is save the test parameters to a settings.settings file from test parameter dialog when user input test parameters so it is available there as well. Then I saved the test parameters to the Nunit30Settings.xml using the same way as how TC saves recent files to the Nunit30Settings.xml. Then added the test parameters to the model by reading it from Nunit30Settings.xml. 
Yes I am building TC from source.

**Comment by CharliePoole** [comment]
Since the settings are global, the approach you're taking is only workable if you never work on any other projects. I think we want the settings to be saved for each individual package we load, similar to what we do for the visual state. Issue #3 is aimed at doing that. I think we need to take issue #3 and expand in more detail.

I looked at your GitHub profile and I see you've done some open source contributing already. I'm looking for folks to join this project. If you're interested, lets start with this general problem. It's a fairly big one, but I think we can divide it into manageable pieces. It's getting a little late, but I'll look at #3 tomorrow and try to lay out some steps I think are needed.

Where are you in the world? I'm near Seattle in the US and on Pacific Daylight Time.

**Comment by YashUoM** [comment]
At the moment I need this for a requirement from my office and making it work for one project is enough for the moment. For that I used the same way TC reloads after adding test parameters using "ChangePackageSettingAndReload("TestParametersDictionary", value);". However I noticed that the commandhandler event is not invoking when I do that, the way it happens on save button click of the testParameterDialog. If I can solve it for now and make it work for the current requirement for my client, it will be great. 

I'm from Sri Lanka on India Standard Time (GMT +5.30)

**Comment by CharliePoole** [comment]
OK... if you need it soon then that's a good approach in the short term.
I'm still inviting you to volunteer on the project for a longer term fix.
:-)

I'm ready to sleep now, but I'll take a look at the CommandHandler tomorrow.

Charlie

**Comment by YashUoM** [comment]
Yes I'd love to contribute for the long term fix. Once I'm done with this I will join it. 
Great. Hope to hear from you soon. 

Yashoda

**Comment by CharliePoole** [comment]
Hi Yashoda,

It looks to me like you ought to modify AppEntry.cs to add parameters to PackageOverrides if present. In that way, you are simulating the presence of a command-line option, which sets the parameters and the model will use the setting each time it creates a new TestPackage.

In fact, an actual parameter isn't such a bad idea, since it would allow scripting execution of the GUI with various parametrs. But just adding it for your own copy will work in the short term.

Charlie

**Comment by YashUoM** [comment]
Yes I did add the parameters to the PackageOverrides. But I havent thought about modifying the AppEntry.cs I will try that and see where it goes, since ChangePackageSettingAndReload("TestParametersDictionary", value) not properly loading the tests after checking testParameters.

**Comment by CharliePoole** [comment]
So the way I usually work these things is in a sequence...

1. Does the change happen correctly after I modify the dialog?

2. Does the change persist when I do subsequent runs?

3. Does the change persist when I do a reload?

4. Does the change persist after I close and restart the application.

It's rare that one of the higher numbered items works correctly unless all
the lower ones work first.

Where are you in that sequenc?

Charlie

**Comment by CharliePoole** [comment]
And, of course, changing AppEntry is intended to deal with the fourth step.

:-)

**Comment by YashUoM** [comment]
Yes I'm in the 4th step but now I've realized that I'm adding the params to PackageOverrides at the wrong step of the initialization. It is too late when I do that. So I hope changing that will work. :)

**Comment by CharliePoole** [comment]
And at that point, you should do no more than add them just as is done for
the other command-line options... don't call ChangePackageSettingsAndReload
since nothing has been loaded as yet.

**Comment by YashUoM** [comment]
Hi Charlie, 

Well I did added the params to  PackageOverrides when creating the model so it is available for the view. However the treeview is not created using that. Any tip on where I can create/recreate the treeview using the test parameters?

**Comment by CharliePoole** [comment]
I don't quite understand... How should the parameters impact the treeview?

**Comment by YashUoM** [comment]
Well, what my test parameter does is get the list of reports which we test. So we need the test tree to be created with as in this image 
![tempsnip](https://user-images.githubusercontent.com/36525289/118442967-edab9b00-b708-11eb-9e16-a6067c3f0d12.png)

This happens when I save test parameters through test parameter dialog. However I'm unable to create the test tree with the same value passed to the view. That's why I tried ChangePackageSettingsAndReload since Its how it happens when loaded via tests parameter dialog

**Comment by CharliePoole** [comment]
Hi Yashoda,

Do you remember I asked you what kind of parameters you were using? That
was because test run parameters, specified at the command line and
retrieved from the TestContext are completely different from method
parameters. But I didn't realize you are trying to use test run parameters
in order to set up run method parameters. This sets up a circularity that
can only work part of the time.

The reason is that TestRunParamters are available only after the test
starts to run. But the method parameters are created  when the test is
discovered, long (I used to say, centuries in computer time) before the
test starts to run... before there is even a test, which could run. Of
course, the parameters could have been designed to work the way you are
using them... they just weren't!

Still, this may be able to work for you in limited conditions. My guess is
that you will have to always use the same names and only change the values
and you will have to reload the tests each time.

Can you send me a copy of AppEntry.cs as you have changed it? I'll take a
look tomorrow.

Charlie

**Comment by YashUoM** [comment]
Hi Charlie, 
We decided not to do this change. Thank you very much for you help. :)

**Comment by CharliePoole** [comment]
OK... let me know if you have more questions.
